src/GeniusYield/TxBuilder/Class.hs

```diff
@@ -77,213 +95,286 @@ module GeniusYield.TxBuilder.Class
 import           GHC.Stack                    (withFrozenCallStack)
 import qualified PlutusLedgerApi.V1           as Plutus (Address, DatumHash,
                                                          FromData (..),
                                                          PubKeyHash, TokenName,
                                                          TxOutRef, Value)
 import qualified PlutusLedgerApi.V1.Value     as Plutus (AssetClass)
 
+-- | Class of monads for building transactions. This can be default derived if the requirements are met.
+-- Specifically, set 'TxBuilderStrategy' to 'GYCoinSelectionStrategy' if you wish to use the default in-house
+-- transaction building implementation.
+class (GYTxSpecialQueryMonad m, GYTxUserQueryMonad m, Default (TxBuilderStrategy m)) => GYTxBuilderMonad m where
+    type TxBuilderStrategy m :: Type
+    type TxBuilderStrategy m = GYCoinSelectionStrategy
+    buildTxBodyWithStrategy :: forall v. TxBuilderStrategy m -> GYTxSkeleton v -> m GYTxBody
+    default buildTxBodyWithStrategy :: forall v. (MonadRandom m, TxBuilderStrategy m ~ GYCoinSelectionStrategy)
+                                    => TxBuilderStrategy m
+                                    -> GYTxSkeleton v
+                                    -> m GYTxBody
+    buildTxBodyWithStrategy = buildTxBodyWithStrategy'
+
+buildTxBody :: GYTxBuilderMonad m => GYTxSkeleton v -> m GYTxBody
+buildTxBody = buildTxBodyWithStrategy def
+
+-- | Class of monads for interacting with the blockchain using transactions.
+class GYTxBuilderMonad m => GYTxMonad m where
+    -- | Sign a transaction body with the user payment key to produce a transaction with witnesses.
+    --
+    -- /Note:/ The key is not meant to be exposed to the monad, so it is only held
+    -- within the closure that signs a given transaction.
+    -- It is recommended to use 'signGYTxBody' and similar to implement this method.
+    signTxBody :: GYTxBody -> m GYTx
+
+    -- | Sign a transaction body with the user payment key AND user stake key to produce
+    -- a transaction with witnesses.
+    -- If the user wallet does not have a stake key, this function should be equivalent to
+    -- 'signTxBody'.
+    --
+    -- See note on 'signTxBody'
+    signTxBodyWithStake :: GYTxBody -> m GYTx
+
+    -- | Submit a fully built transaction to the chain.
+    --   Use 'buildTxBody' to build a transaction body, and 'signGYTxBody' to
+    --   sign it before submitting.
+    --
+    -- /Note:/ Changes made to the chain by the submitted transaction may not be reflected immediately,
+    -- see 'awaitTxConfirmed'.
+    --
+    -- /Law:/ 'someUTxO' calls made after a call to 'submitTx' may return previously returned UTxOs
+    -- if they were not affected by the submitted transaction.
+    submitTx :: GYTx -> m GYTxId
+
+    -- | Wait for a _recently_ submitted transaction to be confirmed.
+    --
+    -- /Note:/ If used on a transaction submitted long ago, the behavior is undefined.
```

Comment:
If it's behaviour is well defined, this can be part of query monad.

Reply:
I like that idea! This should definitely be part of the query monad because yes, it is indeed just a query.

I'd like to wait for the behavior to be defined first, indeed.